<a href="https://colab.research.google.com/github/sankarvinayak/RL-assignments/blob/main/CS24M041_DA6400_Tutorial_6_Actor_Critic.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

# Tutorial: Actor Critic Implementation


In [1]:
# Import required libraries

import argparse
import gymnasium as gym
import numpy as np
from collections import namedtuple

import torch
import torch.nn as nn
import torch.nn.functional as F
import torch.optim as optim
from torch.distributions import Categorical

In [2]:
# Set constants for training # DO NOT CHANGE
seed = 543
log_interval = 10
gamma = 0.99

env = gym.make('CartPole-v1')
env.reset(seed=seed)
torch.manual_seed(seed)

SavedAction = namedtuple('SavedAction', ['log_prob', 'value'])

In [3]:

class Policy(nn.Module):
    """
    implements both actor and critic in one model
    """

    def __init__(self):
        super(Policy, self).__init__()
        self.affine1 = nn.Linear(4, 128)

        # actor's layer
        self.action_head = nn.Linear(128, 2)

        # critic's layer
        self.value_head = nn.Linear(128, 1)

        # action & reward buffer
        self.saved_actions = []
        self.rewards = []

    def forward(self, x):
        """
        forward of both actor and critic
        """
        x = F.relu(self.affine1(x))

        # actor: choses action to take from state s_t
        # by returning probability of each action
        action_prob = F.softmax(self.action_head(x), dim=-1)

        # critic: evaluates being in the state s_t
        state_values = self.value_head(x)

        # return values for both actor and critic as a tuple of 2 values:
        # 1. a list with the probability of each action over the action space
        # 2. the value from state s_t
        return action_prob, state_values

In [4]:
model = Policy()
optimizer = optim.Adam(model.parameters(), lr=3e-2)
eps = np.finfo(np.float32).eps.item()

In [5]:
# DO NOT Modify Training Code

def select_action(state):
    state = torch.from_numpy(state).float()
    probs, state_value = model(state)

    # create a categorical distribution over the list of probabilities of actions
    m = Categorical(probs)

    # and sample an action using the distribution
    action = m.sample()

    # save to action buffer
    model.saved_actions.append(SavedAction(m.log_prob(action), state_value))

    # the action to take (left or right)
    return action.item()


def finish_episode():
    """
    Training code. Calculates actor and critic loss and performs backprop.
    """
    R = 0
    saved_actions = model.saved_actions
    policy_losses = []  # list to save actor (policy) loss
    value_losses = []  # list to save critic (value) loss
    returns = []  # list to save the true values

    # calculate the true value using rewards returned from the environment
    for r in model.rewards[::-1]:
        # calculate the discounted value
        R = r + gamma * R
        returns.insert(0, R)

    returns = torch.tensor(returns)
    returns = (returns - returns.mean()) / (returns.std() + eps)

    for (log_prob, value), R in zip(saved_actions, returns):
        advantage = R - value.item()

        # calculate actor (policy) loss
        policy_losses.append(-log_prob * advantage)

        # calculate critic (value) loss using L1 smooth loss
        value_losses.append(F.smooth_l1_loss(value, torch.tensor([R])))

    # reset gradients
    optimizer.zero_grad()

    # sum up all the values of policy_losses and value_losses
    loss = torch.stack(policy_losses).sum() + torch.stack(value_losses).sum()

    # perform backprop
    loss.backward()
    optimizer.step()

    # reset rewards and action buffer
    del model.rewards[:]
    del model.saved_actions[:]


def train():
    running_reward = 10

    # run infinitely many episodes
    for i_episode in range(2000):

        # reset environment and episode reward
        state = env.reset()[0]
        ep_reward = 0

        # for each episode, only run 9999 steps so that we don't
        # infinite loop while learning
        for t in range(1, 10000):

            # select action from policy
            action = select_action(state)

            # take the action
            state, reward, done, truncated, _ = env.step(action)

            model.rewards.append(reward)
            ep_reward += reward
            if done:
                break

        # update cumulative reward
        running_reward = 0.05 * ep_reward + (1 - 0.05) * running_reward

        # perform backprop
        finish_episode()

        # log results
        if i_episode % log_interval == 0:
            print('Episode {}\tLast reward: {:.2f}\tAverage reward: {:.2f}'.format(
                  i_episode, ep_reward, running_reward))

        # check if we have "solved" the cart pole problem
        if running_reward > env.spec.reward_threshold:
            print("Solved! Running reward is now {} and "
                  "the last episode runs to {} time steps!".format(running_reward, t))
            break

In [6]:
train()

Episode 0	Last reward: 22.00	Average reward: 10.60
Episode 10	Last reward: 28.00	Average reward: 16.78
Episode 20	Last reward: 42.00	Average reward: 33.66
Episode 30	Last reward: 21.00	Average reward: 31.73
Episode 40	Last reward: 26.00	Average reward: 29.00
Episode 50	Last reward: 150.00	Average reward: 64.74
Episode 60	Last reward: 86.00	Average reward: 88.06
Episode 70	Last reward: 209.00	Average reward: 118.08
Episode 80	Last reward: 63.00	Average reward: 169.06
Episode 90	Last reward: 36.00	Average reward: 159.00
Episode 100	Last reward: 199.00	Average reward: 175.59
Episode 110	Last reward: 288.00	Average reward: 209.42
Episode 120	Last reward: 268.00	Average reward: 260.10
Episode 130	Last reward: 320.00	Average reward: 399.67
Episode 140	Last reward: 191.00	Average reward: 296.83
Episode 150	Last reward: 184.00	Average reward: 254.95
Episode 160	Last reward: 183.00	Average reward: 208.01
Episode 170	Last reward: 584.00	Average reward: 288.11
Solved! Running reward is now 840.36

# TODO: Write a policy class similar to the above, without using shared features for the actor and critic and compare their performance.


In [7]:

# TODO: Write a policy class similar to the above, without using shared features for the actor and critic and compare their
# performance.

class UnsharedPolicy(nn.Module):
    def __init__(self):
        super(UnsharedPolicy, self).__init__()
        # TODO: Fill in.
        hidden_size = 128
        # Actor network
        self.actor_affine1 =nn.Linear(4,hidden_size)
        self.action_head =nn.Linear(hidden_size,2)

        # Critic network
        self.critic_affine1 =nn.Linear(4,hidden_size)
        self.value_head =nn.Linear(hidden_size,1)

        self.saved_actions = []
        self.rewards = []

    def forward(self, x):
        # TODO: Fill in. For your networks, use the same hidden_size for the layers as the previous policy, that is 128.
        # Actor forward pass
        actor_x =F.relu(self.actor_affine1(x))
        action_prob =F.softmax(self.action_head(actor_x), dim=-1)

        # Critic forward pass
        critic_x =F.relu(self.critic_affine1(x))
        state_values =self.value_head(critic_x)
        # return values for both actor and critic as a tuple of 2 values:
        # 1. A list with the probability of each action over the action space
        # 2. The value from state s_t
        return action_prob, state_values



In [8]:
%%time
model = UnsharedPolicy()
optimizer = optim.Adam(model.parameters(), lr=3e-2)
eps = np.finfo(np.float32).eps.item()
train()

Episode 0	Last reward: 88.00	Average reward: 13.90
Episode 10	Last reward: 13.00	Average reward: 13.18
Episode 20	Last reward: 9.00	Average reward: 11.88
Episode 30	Last reward: 9.00	Average reward: 10.93
Episode 40	Last reward: 10.00	Average reward: 10.26
Episode 50	Last reward: 10.00	Average reward: 9.95
Episode 60	Last reward: 9.00	Average reward: 9.59
Episode 70	Last reward: 10.00	Average reward: 9.46
Episode 80	Last reward: 10.00	Average reward: 9.43
Episode 90	Last reward: 9.00	Average reward: 9.44
Episode 100	Last reward: 9.00	Average reward: 9.25
Episode 110	Last reward: 10.00	Average reward: 9.40
Episode 120	Last reward: 10.00	Average reward: 9.57
Episode 130	Last reward: 10.00	Average reward: 9.72
Episode 140	Last reward: 48.00	Average reward: 16.60
Episode 150	Last reward: 49.00	Average reward: 35.68
Episode 160	Last reward: 108.00	Average reward: 48.75
Episode 170	Last reward: 57.00	Average reward: 51.77
Episode 180	Last reward: 50.00	Average reward: 61.49
Episode 190	Last 

In [14]:
weight_difference = torch.norm(model.actor_affine1.weight - model.critic_affine1.weight)
print("Weight difference:", weight_difference.item())

# for a trained netowrk difference in weights of actor and critic

Weight difference: 20.67447280883789


In [9]:
x=torch.rand(1,4)

In [12]:
actor=model.actor_affine1(x)
critic=model.critic_affine1(x)

In [13]:
F.cosine_similarity(actor, critic, dim=-1)

#cosine similarity of the hidden repreentaiton of state between actor and critic for a random value input

tensor([0.1842], grad_fn=<SumBackward1>)

In [22]:
np.array([1,2,3,4])

array([1, 2, 3, 4])

In [23]:
env.reset()

(array([-0.03684774, -0.03309242,  0.02791581,  0.03389481], dtype=float32),
 {})

In [24]:
x,_=env.reset()
x=torch.from_numpy(x).float()

In [25]:
actor=model.actor_affine1(x)
critic=model.critic_affine1(x)
F.cosine_similarity(actor, critic, dim=-1)
#cosine similarity of the hidden repreentaiton of state between actor and critic for a value in the valid range as input

tensor(0.4685, grad_fn=<SumBackward1>)

In [15]:
model=UnsharedPolicy()
weight_difference = torch.norm(model.actor_affine1.weight - model.critic_affine1.weight)
print("Weight difference:", weight_difference.item())
# for a randomly initialized netowrk difference in weights of actor and critic

Weight difference: 9.285624504089355


## Observarions
- The initial values of single network is high compared to that of two netowork
- There is much more varience in the last reward obtained in single network which may due to the double update performed one from policy and one from value
- The averge reward of Unshared network initially decrease and then increase which may be due to the changing of policy and value during traing not sharing the prameters
- The shared network the average reward is increasing almost monotolically as the actor and critic are getting updated by each other
- It is not necessasy that is representations learned in the netowkr are similar for actor and critic as the cosine similarity for a random value for a random vector is 0.18 which is queit low and for a value in range of environment of 0.46 shows moderate corelation between the representation of both netowrok